thunder/transforms/constant_folding.py

```diff
@@ -70,6 +76,89 @@ def materialize_args(a):
 
 
 class ConstantFolding(thunder.Transform):
+    """Apply Constant Folding to computation trace.
+
+    With this transform applied to a computation trace, successive passes
+    (meaning trace transformations) can transform the simplified compute.
+
+
+    .. code-block:: python
+        :name: example-constant_folding
+
+        from thunder.transforms import ConstantFolding
+
+        model = ...
+        transforms = [ConstantFolding()]
+        jitted = thunder.jit(model, transforms=transforms)
+        # If you prefer `ThunderCompiler`...
+        from thunder.dynamo import ThunderCompiler
+        backend = ThunderCompiler(transforms=transforms)
+        jitted = torch.compile(model, backend=backend)
+
+
+    To see the effect of this transform, let's use the following function:
+
+    .. code-block:: python
+
+        def forward(x):
+            scale_t = torch.tensor([2.])
+            getitem = scale_t[0]
+            getitem_2 = torch.tensor([2., 3.])[0]
+            return x + getitem + getitem_2
```

Comment:
Maybe we can use an example which will generate smaller traces (and easily highlight ConstantFolding)?

Eg.
```python
def forward(x):
    scale_t = torch.tensor([2.])
    scale_t = (scale_t * 10) / 5
    return x * scale_t
```

Initial Trace
```python
def computation(x):
  # x: "cpu f32[3]"

  scale_t = ltorch.tensor([2.0], device=None, dtype=None, requires_grad=False, pin_memory=False)  # scale_t: "cpu f32[1]"
    # scale_t = prims.tensor_from_sequence([2.0], dtype=None, device=devices.Device("cpu"))  # scale_t: "cpu f32[1]"

  t1 = ltorch.mul(scale_t, 10)  # t1: "cpu f32[1]"
    # _ = prims.convert_element_type(10, float)
    # t1 = prims.mul(scale_t, 10.0)  # t1: "cpu f32[1]"
  t2 = ltorch.true_divide(t1, 5)  # t2: "cpu f32[1]"
    # _ = prims.convert_element_type(5, float)
    # t2 = prims.div(t1, 5.0)  # t2: "cpu f32[1]"

  t4 = ltorch.mul(x, t2)  # t4: "cpu f32[3]"
    # t3 = prims.broadcast_in_dim(t2, (3,), (0,))  # t3: "cpu f32[3]"
    # t4 = prims.mul(x, t3)  # t4: "cpu f32[3]"
  return t4
```

Const Folded Trace
```python
def computation(x):
  # x: "cpu f32[3]"
  t2 = prims.tensor_from_sequence([4.0], dtype=dtypes.float32, device=devices.Device("cpu"))  # t2: "cpu f32[1]"

  t4 = ltorch.mul(x, t2)  # t4: "cpu f32[3]"
    # t3 = prims.broadcast_in_dim(t2, (3,), (0,))  # t3: "cpu f32[3]"
    # t4 = prims.mul(x, t3)  # t4: "cpu f32[3]"
  return {'output': t4, 'flat_args': [x]}
```

Reply:
@crcrpar Thank you for doing this! Do you want move to the shorter example or merge as is?

Reply:
copied Kshiteej's example.